docs/design/build-system.md

```diff
@@ -0,0 +1,154 @@
+# Build System
+
+Angular CLI includes a first-party build system for Angular applications distributed as `@angular-devkit/build-angular`.
+This build system is responsible for creating a standalone application from user source files and third-party dependencies.
+
+`@angular-devkit/build-angular` itself integrates with the rest of Angular CLI by being an [Architect builder](https://angular.io/guide/cli-builder).
+This document describes a top level view of the functionality in `@angular-devkit/build-angular`, referred as just the "build system".
+Deprecated or soon to be removed features are not described here.
+
+In broad strokes the main areas are:
+- loading and processing sources
+- code splitting
+- production optimizations
+- post-processing steps
+
+Many tools are used in this process, and most of these steps happen within a [Webpack](https://webpack.js.org/) build.
+We maintain a number of Webpack-centric plugins in this repository, some of these are public but most are private since they are very specific to our setup.
+
+
+## Loading and processing sources
+
+Sources for Angular CLI browser apps are comprised of TypeScript files, style sheets, assets, scripts, and third party dependencies.
+A given build will load these sources from disk, process them, and bundle them together.
+
+
+### TypeScript
+
+Angular builds rely heavily on TypeScript-specific functionality for Ahead-of-Time template compilation (AOT).
+Outside Angular CLI, this is performed by the Angular Compiler (`ngc`), provided by `@angular/compiler-cli`.
+To avail of Ahead-of-Time template compilation within a Webpack compilation we use and distribute the `@ngtools/webpack` Webpack plugin.
+
+Typescript sources are loaded from disk and compiled in-memory into Javascript files that are stored in a virtual file system and made available to Webpack.
+During compilation we also perform a number of code transformations using TypeScript transformers that enable automatic usage of AOT, internationalization features, and server-side rendering.
+
+AOT compilation requires loading HTML and CSS resources referenced on Angular Components. These are processed using a Webpack child compilation.
+
+The build system allows specifying replacements for specific files by replacing what path is loaded from the virtual file system.
+This is used for conditional loading of code at build time.
+
+### Stylesheets
+
+Two types of stylesheets are used in the build system: global stylesheets and component stylesheets.
+Global stylesheets are injected into the `index.html` file, while component stylesheets are loaded directly into compiled Angular components.
```

Comment:
Are global stylesheets also injected into other entry points? Or are all Angular apps actual SPA's under the hood?

Reply:
`index.html` is usually the only entry-point per se. I suppose the real answer is that the CLI build system only produces SPAs, yes. The server builder makes a universal app but at that point you are already in control of what's served.

Reply:
Thanks for the clarification. It might be worth pointing out somewhere that the CLI only builds SPAs (doesn't necessarily need to be done in the stylesheets section).